packages/Search/src/details/DTK_DetailsTreeConstruction_decl.hpp

```diff
@@ -95,6 +101,169 @@ struct TreeConstruction
     static Kokkos::pair<int, int> determineRange(
         Kokkos::View<unsigned int *, DeviceType> sorted_morton_codes, int i );
 };
+
+template <typename ViewType>
+class CalculateBoundingBoxOfTheSceneFunctor
+{
+  public:
+    CalculateBoundingBoxOfTheSceneFunctor(
+        typename ViewType::const_type primitives )
+        : _primitives( primitives )
+    {
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void init( Box &box ) const { box = Box(); }
+
+    KOKKOS_INLINE_FUNCTION
+    void operator()( int const i, Box &box ) const
+    {
+        expand( box, _primitives( i ) );
+    }
+
+    KOKKOS_INLINE_FUNCTION
+    void join( volatile Box &dst, volatile Box const &src ) const
+    {
+        expand( dst, src );
+    }
+
+  private:
+    typename ViewType::const_type _primitives;
+};
+
+template <typename DeviceType>
+template <typename ConstViewType>
+inline void TreeConstruction<DeviceType>::calculateBoundingBoxOfTheScene(
+    ConstViewType primitives, Box &scene_bounding_box )
+{
+    static_assert( Kokkos::is_view<ConstViewType>::value, "Must pass a view" );
+    static_assert( std::is_same<typename ConstViewType::traits::device_type,
+                                DeviceType>::value,
+                   "Wrong device type" );
+    // TODO static_assert( is_expandable_v<Box, typename
+    // ConstViewType::value_type), "");
+    auto const n = primitives.extent( 0 );
+    Kokkos::parallel_reduce(
+        DTK_MARK_REGION( "calculate_bounding_box_of_the_scene" ),
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n ),
+        CalculateBoundingBoxOfTheSceneFunctor<decltype( primitives )>(
+            primitives ),
+        scene_bounding_box );
+    Kokkos::fence();
+}
+
+template <typename Primitives, typename MortonCodes>
+inline void assignMortonCodesDispatch( BoxTag, Primitives primitives,
+                                       MortonCodes morton_codes,
+                                       Box const &scene_bounding_box )
+{
+    using ExecutionSpace =
+        typename decltype( primitives )::traits::execution_space;
+    auto const n = morton_codes.extent( 0 );
+    Kokkos::parallel_for(
+        DTK_MARK_REGION( "assign_morton_codes" ),
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n ), KOKKOS_LAMBDA( int i ) {
+            Point xyz;
+            centroid( primitives( i ), xyz );
```

Comment:
These two lines almost call for having a function to do `xyz = centroid(primitives(i))`.

Reply:
This is what `returnCentroid()` does.  Haven't had a chance to play with the compiler explorer if that has any effect when the optimizer is turned on.

Reply:
I am tempted to leave it this way because it will need to be revisited when I define what is required to have a specialization in the traits for a user input to be valid.